workers/unity/Packages/com.improbable.gdk.core/Subscriptions/Systems/CommandTaskSystem.cs

```diff
@@ -0,0 +1,45 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Improbable.Gdk.Core.Commands;
+using Unity.Entities;
+
+namespace Improbable.Gdk.Subscriptions
+{
+    [DisableAutoCreation]
+    public class CommandTaskSystem : ComponentSystem
+    {
+        private readonly Dictionary<Type, ITaskManager> taskManagers =
+            new Dictionary<Type, ITaskManager>();
+
+        public void RegisterTask<T>(long requestId, TaskCompletionSource<T> task)
```

Comment:
It might be preferable to have this return a Task for a given ID. That way you can hide that it's actually a `TaskCompletionSource` that is being used. If at all possible, present a user with a basic `Task`.

Reply:
Basically I think this might make more sense as an async command sender, as that can be used for more than just the subscriptions stuff. However for the purposes of this exercise it's not really important.

Reply:
The remaining issue with that is the lifecycle stuff. But that should be fixed by making it so that subscribed information is published at the right time. Rather than them registering with some central thing. (Not for this PR, it's a mistake in how subscriptions is currently setup).

Reply:
(Which is why the non-constraint callback managers are also setup like this)

Reply:
oh yeah, I like the idea of hiding the `TaskCompletionSource` and agree that the lifecycle system, while convenient, makes it a bit more tricky to add custom functionality to it.